heartbeat/scheduler/timerqueue/queue.go

```diff
@@ -0,0 +1,163 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timerqueue
+
+import (
+	"container/heap"
+	"context"
+	"time"
+)
+
+// TimerTask represents a task run by the TimerQueue.
+type TimerTask struct {
+	fn    TimerTaskFn
+	runAt time.Time
+}
+
+// NewTimerTask creates a new TimerTask struct.
+func NewTimerTask(runAt time.Time, fn TimerTaskFn) *TimerTask {
+	return &TimerTask{runAt: runAt, fn: fn}
+}
+
+// TimerTaskFn is the function invoked by a TimerTask.
+type TimerTaskFn func(now time.Time)
+
+// TimerQueue represents a priority queue of timers.
+type TimerQueue struct {
+	th            *timerHeap
+	ctx           context.Context
+	nextRunAt     *time.Time
+	popRunnableCh chan chan []*TimerTask
+	pushCh        chan *TimerTask
+	timer         *time.Timer
+}
+
+// NewTimerQueue creates a new instance.
+func NewTimerQueue(ctx context.Context) *TimerQueue {
+	tq := &TimerQueue{
+		th:            &timerHeap{},
+		ctx:           ctx,
+		popRunnableCh: make(chan chan []*TimerTask),
+		pushCh:        make(chan *TimerTask),
+		timer:         time.NewTimer(0),
+	}
+	heap.Init(tq.th)
+
+	return tq
+}
+
+// Push adds a task to the queue
+func (tq *TimerQueue) Push(tt *TimerTask) {
```

Comment:
hm, do we even need to expose the TimerTask type? How about `func (tq *TimerQueue) Push(runAt time.Time, fn TimerTaskFn)` ?

Reply:
Good point, we don't!